src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+    return r;
+}
+
+SOL_API int
+sol_http_create_simple_uri(char **uri, const char *base_url,
+    const struct sol_http_param *params)
+{
+    char *encoded_params;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+    SOL_NULL_CHECK(base_url, -EINVAL);
+
+    if (params) {
+        r = sol_http_encode_params(&encoded_params, SOL_HTTP_PARAM_QUERY_PARAM,
+            params);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    if (asprintf(uri, "%s/%s%s", base_url, *encoded_params ? "?" : "",
+        encoded_params) < 0) {
+        r = -ENOMEM;
+        SOL_ERR("Could not create the URL with the encoded parameters."
+            "URL: %s, parameters: %s", base_url, encoded_params);
+    }
+    free(encoded_params);
+    return r;
+}
+
+static bool
+_is_sub_delimin(char c)
+{
+    if (c != '!' && c != '$' && c != '&' && c != '\''
+        && c != '(' && c != ')' && c != '*' && c != '+' && c != ','
+        && c != ';' && c != '=')
+        return false;
+    return true;
+}
+
+static bool
+_is_unreserved(char c)
+{
+    if (!isalnum(c) && c != '-' && c != '.' && c != '_' && c != '~')
+        return false;
+    return true;
+}
+
+static bool
+_is_pct_encoded(char c)
+{
+    if (c != '%')
+        return false;
+    return true;
+}
+
+static bool
+_is_pchar(char c)
+{
+    if (!_is_unreserved(c) && !_is_pct_encoded(c) &&
+        !_is_sub_delimin(c) && c != ':' && c != '@')
+        return false;
+    return true;
+}
+
+static int
+_get_protocol(const char *uri, struct sol_str_slice *protocol, char **end)
+{
+    struct sol_str_slice p;
+    size_t i;
+
+    p.data = uri;
+    p.len = 0;
+
+    for (i = 0; uri[i] && uri[i] != ':'; i++, p.len++) {
+        if (!isalnum(uri[i]) && uri[i] != '+' &&
+            uri[i] != '-' && uri[i] != '.') {
+            SOL_ERR("Invalid character at the URI's scheme: %c", uri[i]);
+            return -EINVAL;
+        }
+    }
+
+    if (uri[i] != ':') {
+        SOL_ERR("Could not find the \":\" scheme");
+        return -EINVAL;
+    }
+
+    SOL_DBG("URI Protocol: %.*s", SOL_STR_SLICE_PRINT(p));
+    *end = (char *)p.data + p.len + 1;
+
+    if (protocol)
+        *protocol = p;
+
+    return 0;
+}
+
+static int
+_get_server(char *partial_uri, struct sol_str_slice *server, char **next)
+{
+    char *i;
+    struct sol_str_slice s;
+    size_t j;
+
+    for (j = 0; j < 2; j++, partial_uri++) {
+
+        if (!*partial_uri) {
+            SOL_ERR("Expecting a \"/\" character to parse the server. "
+                "Found: NUL");
+            return -EINVAL;
+        }
+
+        if (*partial_uri != '/') {
+            SOL_ERR("Expecting a \"/\" character to parse the server. "
+                "Found: %c", *partial_uri);
+            return -EINVAL;
+        }
+    }
+
+    if (!*partial_uri) {
+        SOL_ERR("Missing host information.");
+        return -EINVAL;
+    }
+
+    s.data = partial_uri;
+
+    for (i = partial_uri; *i && *i != '/' && *i != ':' && *i != '#'
+        && *i != '?'; i++) {
```

Comment:
readability hint: use `for ()` with the main condition you're walking, like `for (i = partial_uri; *i; i++)` or `for (idx = 0; idx < partial_uri.len; idx++)`. Then inside the for you check for the other stop condition.

It will be easier to spot problems and to review.